rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,315 @@
+Vrl doesn't search on columns that do not exist within the datafile. Searching
+on a column that doesn't exist can prevent Vector from loading.
+
+If (when these features are implemented) the user attempts to reload a Resource
+and a column that has been Indexed no longer exists, this should prevent the
+file from being loaded. Vector will continue to use the currently loaded data.
+
+These restrictions apply only to columns that VRL is searching on. If the user
+attempts to use a field that has not been returned from the enrichment, this
+will be possible, but the value will be Null. This is largely due to the way
+that VRL works with paths into Objects.
+
+### Vrl functions
+
+Two remap functions:
+
+#### `find_row`
+
+This function will look up a single row within the dataset. If a single row is
+found that data is returned as an object, otherwise this function will error.
+
+A metric will be emitted to indicate the lookup time.
+
+#### `search_rows`
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the
+array.
+
+A metric will be emitted to indicate the lookup time.
+
+#### Parameters
+
+*resource*
+
+The name of the resource to lookup. This must point to a resource specified
+in the config file eg `resources.csv_file`.
+
+*condition*
+
+`condition` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
+
+*case_sensitive*
+
+By default the search will be case insensitive, but this can be changed by
+passing `true` to this parameter.
+
+### Example config
+
+```toml
+[resources.csv]
+    type = "csv"
+    file = "/path/to/csv.csv"
+    delimiter = ","
+
+[sources.datadog_logs]
+    type = "datadog_logs"
+    address = "0.0.0.0:80"
+
+[transforms.simple_enrich]
+    type = "remap"
+    inputs = ["datadog_logs"]
+    source = '''
+        . = parse_json!(.message)
+
+        result, err = find_row(
+            resources.csv,
+            { "license_plate": .license }
+        )
+
+        if is_null(err) {
+            .first_name = result.first_name
+            .last_name = result.last_name
+        }
+   '''
+```
+
+## Internal Proposal
+
+We will need to add a new component type to Vector, call it `Resource`. On
+loading the Vector config these instances will be created and will load the
+data that they are pointing to.
+
+The entire data file will be loaded into memory, so all lookups will be
+performed in memory. A `Resource` will need to provide threadsafe,
+readonly access to the data that it loads.
+
+VRL will need to maintain the concept of `Resource`. This can be created as
+an additional element to the `vrl::Value` type. On compilation, the available
+resources can be added to the Variable type definitions. This ensures that
+during compilation the functions will access valid resources.
+
+### Indexing
```

Comment:
Actually, you note in the "scope" section that this is out of scope. So should we remove this section altogether?

Reply:
I think it is worth keeping this in. I would like to keep the restriction to ensure we know the search fields at compile time for the first version. It wouldn't strictly be necessary without indexing, but since we are anticipating indexing this helps to explain the restriction.